File: permissionsex-sponge/src/main/java/ninja/leaping/permissionsex/sponge/PEXPermissionDescription.java
/**
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.leaping.permissionsex.sponge;

import com.google.common.base.Preconditions;
import ninja.leaping.permissionsex.util.Tristate;
import org.spongepowered.api.plugin.PluginContainer;
import org.spongepowered.api.service.permission.PermissionDescription;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.text.Text;

import java.util.HashMap;
import java.util.Map;

/**
 * Implementation of a permission description
 */
class PEXPermissionDescription implements PermissionDescription {
    private final PermissionsExPlugin plugin;
    private final String permId;
    private final Text description;
    private final PluginContainer owner;

    public PEXPermissionDescription(PermissionsExPlugin plugin, String permId, Text description, PluginContainer owner) {
        this.plugin = plugin;
        this.permId = permId;
        this.description = description;
        this.owner = owner;
    }

    @Override
    public String getId() {
        return this.permId;
    }

    @Override
    public Text getDescription() {
        return this.description;
    }

    @Override
    public Map<Subject, Boolean> getAssignedSubjects(String type) {
        return plugin.getSubjects(type).getAllWithPermission(getId());
    }

    @Override
    public PluginContainer getOwner() {
        return this.owner;
    }

    static class Builder implements PermissionDescription.Builder {
        private final PluginContainer owner;
        private final PermissionsExPlugin plugin;
        private String id;
        private Text description;
        private Map<String, Tristate> ranks = new HashMap<>();

        Builder(PluginContainer owner, PermissionsExPlugin plugin) {
            this.owner = owner;
            this.plugin = plugin;
        }

        @Override
        public Builder id(String id) {
            Preconditions.checkNotNull(id, "id");
            this.id = id;
            return this;
        }

        @Override
        public Builder description(Text text) {
            Preconditions.checkNotNull(text, "text");
            this.description = text;
            return this;
        }

        @Override
        public Builder assign(String s, boolean b) {
            return assign(s, b ? Tristate.TRUE : Tristate.FALSE);
        }

        public Builder assign(String rankTemplate, Tristate power) {
            ranks.put(rankTemplate, power);
            return this;
        }

        @Override
        public PEXPermissionDescription register() throws IllegalStateException {
            Preconditions.checkNotNull(id, "id");
            Preconditions.checkNotNull(description, "description");

            final PEXPermissionDescription ret = new PEXPermissionDescription(plugin, this.id, this.description, this.owner);
            this.plugin.registerDescription(ret, ranks);
            return ret;
        }
    }
}

File: platform/sponge7/src/main/kotlin/ca/stellardrift/permissionsex/sponge/PEXPermissionDescription.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ca.stellardrift.permissionsex.sponge

import java.util.HashMap
import java.util.Optional
import java.util.concurrent.CompletableFuture
import org.spongepowered.api.plugin.PluginContainer
import org.spongepowered.api.service.permission.PermissionDescription
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.text.Text

/**
 * Implementation of a permission description
 */
class PEXPermissionDescription(
    private val plugin: PermissionsExPlugin,
    private val permId: String,
    description: Text?,
    owner: PluginContainer?
) : PermissionDescription {
    private val description: Optional<Text> = description.optionally()
    private val owner: Optional<PluginContainer> = owner.optionally()

    override fun getId(): String {
        return this.permId
    }

    override fun getDescription(): Optional<Text> {
        return this.description
    }

    override fun findAssignedSubjects(type: String): CompletableFuture<Map<SubjectReference, Boolean>> {
        return this.plugin.loadCollection(type).thenCompose { it.getAllWithPermission(id) }
    }

    override fun getAssignedSubjects(collectionIdentifier: String): Map<Subject, Boolean> {
        return this.plugin.getCollection(collectionIdentifier).map { it.getLoadedWithPermission(id) }.orElseGet { emptyMap() }
    }

    override fun getOwner(): Optional<PluginContainer> {
        return this.owner
    }

    internal class Builder(private val owner: PluginContainer, private val plugin: PermissionsExPlugin) : PermissionDescription.Builder {
        private var id: String? = null
        private var description: Text? = null
        private val ranks: MutableMap<String, Int> = HashMap()

        override fun id(id: String): Builder {
            this.id = id
            return this
        }

        override fun description(text: Text?): Builder {
            this.description = text
            return this
        }

        override fun assign(s: String, b: Boolean): Builder {
            return assign(s, if (b) 1 else -1)
        }

        fun assign(rankTemplate: String, power: Int): Builder {
            this.ranks[rankTemplate] = power
            return this
        }

        @Throws(IllegalStateException::class)
        override fun register(): PEXPermissionDescription {
            val ret = PEXPermissionDescription(this.plugin, this.id!!, this.description, this.owner)
            this.plugin.registerDescription(ret, this.ranks)
            return ret
        }
    }
}
